py/bitbox02/bitbox02/communication/bitbox_api_protocol.py

```diff
@@ -404,6 +425,66 @@ def _handshake_query(self, req: bytes) -> Tuple[bytes, bytes]:
     def _decode_noise_response(self, encrypted_msg: bytes) -> Tuple[bytes, bytes]:
         return encrypted_msg[:1], encrypted_msg[1:]
 
+    # Send the message to the device, resend if the device is busy.
+    # Poll for a result until the query is completed.
+    def _execute_raw_query(self, msg: bytes, cid: int) -> bytes:
+        status = None
+        payload = None
+        msg_sent = False
+        while not msg_sent:
+            response = self._transport.query(HwwRequestCode.HWW_REQ_NEW + msg, HWW_CMD, cid)
+            assert len(response) != 0, "Unexpected response of length 0 from HWW stack."
+            status, payload = response[0:1], response[1:]
+            if status != HwwResponseCode.HWW_RSP_BUSY:
+                msg_sent = True
+            else:
+                assert (
+                    len(payload) == 0
+                ), "Unexpected payload of length {} with HWW_RSP_BUSY response.".format(
+                    len(payload)
+                )
+
+        if status in [HwwResponseCode.HWW_RSP_NACK]:
+            # We should never receive a NACK unless some internal error occurs.
+            raise Exception("Unexpected NACK response from HWW stack.")
+
+        # The message has been sent. If we have a retry, poll the device until we're ready.
+        while status == HwwResponseCode.HWW_RSP_NOT_READY:
+            assert (
+                len(payload) == 0
+            ), "Unexpected payload of length {} with HWW_RSP_NOT_READY response.".format(
+                len(payload)
+            )
+            time.sleep(0.2)
```

Comment:
not imported, does this even work? :D

Reply:
Got lost in the process! Fixed in 2756844.